packages/bindings-test/src/multitest.rs

```diff
@@ -100,23 +100,26 @@ impl Pool {
         let (final_in, final_out, payout) = match amount {
             SwapAmount::In(input) => {
                 let input_minus_fee = input * (Decimal::one() - self.fee);
-                let final_out = bal_in * bal_out / (bal_in + input_minus_fee);
-                let payout = SwapAmount::Out(bal_out - final_out);
-                let final_in = bal_in + input;
+                let final_out =
+                    bal_in.checked_mul(bal_out)? / (bal_in.checked_add(input_minus_fee)?);
+                let payout = SwapAmount::Out(bal_out.checked_sub(final_out)?);
+                let final_in = bal_in.checked_add(input)?;
                 (final_in, final_out, payout)
             }
             SwapAmount::Out(output) => {
-                let in_without_fee = bal_in * bal_out / (bal_out - output);
+                let in_without_fee = bal_in.checked_mul(bal_out)? / bal_out.checked_sub(output)?;
```

Comment:
Oh yes, good to use the checked version.
But I guess we could imagine adding even more context to these?

Reply:
I think `thiserror` doesn't allow expanding messages imported by `#[from]` from other crates... So to add more context, I can't think of other solution then to add matchers for every case, catch specific error and return our own with proper context.
That's a lot of cumbersome boilerplate.
Do you have some better idea?

Reply:
My only idea was

`bal_in.checked_mul(bal_out).map_err(with_context(line!()))? / ...`

Where

```rust
use cosmwasm_std::{OverflowError, StdError};
fn with_context(line: u32) -> Box<dyn Fn(OverflowError) -> StdError> {
    let context = move |err: OverflowError| {
        StdError::generic_err(format!("Overflow {} on line {}", err.to_string(), line))
    };
    Box::new(context)
}
```

I am sure you can come up with a cleaner way to make this closure, and yes it is a bit verbose, adding `.map_err(with_context(line!()))` everywhere, but maybe you can make it shorter, like `.map_err(qq(line!()))`

Reply:
A panic handler that turns the panic into an error looks like a great idea. Perhaps it can have different logging modes,.depending on the release / debug builds, by example.

Reply:
The only issue we have with panics is their behaviour in Wasm.

Unit tests give us the exact line:

```
running 1 test
test contract::tests::cosmos_signature_verify_works ... FAILED

failures:

---- contract::tests::cosmos_signature_verify_works stdout ----
thread 'contract::tests::cosmos_signature_verify_works' panicked at 'kaputt', src/contract.rs:99:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    contract::tests::cosmos_signature_verify_works
```

Or backtrace:

```
running 1 test
test contract::tests::cosmos_signature_verify_works ... FAILED

failures:

---- contract::tests::cosmos_signature_verify_works stdout ----
thread 'contract::tests::cosmos_signature_verify_works' panicked at 'kaputt', src/contract.rs:99:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143:14
   2: crypto_verify::contract::query_verify_cosmos
             at ./src/contract.rs:99:5
   3: crypto_verify::contract::query
             at ./src/contract.rs:35:25
   4: crypto_verify::contract::tests::cosmos_signature_verify_works
             at ./src/contract.rs:278:19
   5: crypto_verify::contract::tests::cosmos_signature_verify_works::{{closure}}
             at ./src/contract.rs:265:5
   6: core::ops::function::FnOnce::call_once
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ops/function.rs:227:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Reply:
I spent quite some time on this early in CosmWasm (2020), also trying different panic handling in Cargo.toml (unwind) and even using a catch_unwind handler to make an error message.

nothing worked and I remember that I discovered unwrapping panics was not supported yet.  not sure if it was the wasm spec or the wasmer implementation. 

if you find a way to do so, amazing! but I have a strong feeling this is still impossible. also backtraces don't come out of wasm contracts.

the questions here is, in absence of solving the panic message, how do we get useful info out of a production contract when someone puts in numbers that over/underflow 

"unreachable" is not useful.

"cannot subtract 10 and 5" is more useful but still a bit hard to debug if you have lots of math in the function.

discussion is how to add some more useful info from wasm contracts on a blockchain. the unit test debugging works well

Reply:
@webmaster128 
> I'm not sure that using checked_* math is really the way to go if you don't handle the errors but just abort execution.

How it works currently, with plain panics:
```
---- multitest::repay_with_collateral::on_two_markets stdout ----
thread 'multitest::repay_with_collateral::on_two_markets' panicked at 'attempt to subtract with overflow', /home/u/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmwasm-std-1.0.0-rc.0/src/math/uint128.rs:322:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
with backtrace
```
thread 'multitest::repay_with_collateral::on_two_markets' panicked at 'attempt to subtract with overflow', /home/u/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmwasm-std-1.0.0-rc.0/src/math/uint128.rs:322:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:116:14
   2: core::panicking::panic_display
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:72:5
   3: core::panicking::panic_str
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:56:5
   4: core::option::expect_failed
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/option.rs:1817:5
   5: core::option::Option<T>::expect
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/option.rs:692:21
   6: <cosmwasm_std::math::uint128::Uint128 as core::ops::arith::Sub>::sub
             at /home/u/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmwasm-std-1.0.0-rc.0/src/math/uint128.rs:320:13
   7: osmo_bindings_test::multitest::Pool::swap
             at /home/u/.cargo/registry/src/github.com-1ecc6299db9ec823/osmo-bindings-test-0.5.1/src/multitest.rs:109:57
```
And then I have to either way clone repo, add debug prints and modify my `Cargo.toml` to get actual values that are incorrect.

Now with `checked_` math version I added here:
```
---- multitest::repay_with_collateral::on_two_markets stdout ----
thread 'multitest::repay_with_collateral::on_two_markets' panicked at 'called `Result::unwrap()` on an `Err` value: error executing WasmMsg:
sender: user
Execute { contract_addr: "contract1", msg: Binary(7b22.....d7d7d), funds: [] }

Caused by:
    0: error executing WasmMsg:
       sender: contract1
       Execute { contract_addr: "contract2", msg: Binary(7b22737761705f77697468647261775f66726f6d223a7b226163636f756e74223a2275736572222c2273656c6c5f6c696d6974223a2232303030303030222c22627579223a7b2264656e6f6d223a7b224e6174697665223a22455448227d2c22616d6f756e74223a2231303030303030227d7d7d), funds: [] }
    1: Generic error: Querier contract error: Cannot Sub with 20000 and 557227
    2: Generic error: Querier contract error: Cannot Sub with 20000 and 557227', contracts/isotonic-credit-agency/src/multitest/repay_with_collateral.rs:212:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
it looks better for debugging, I get some actual details - but with no source.
Drawback here - backtrace will get me nowhere, since it's full of unwinding panic (hangling error).